plutus-ir/src/Language/PlutusIR/Compiler/Let.hs

```diff
@@ -0,0 +1,84 @@
+        body' <- compileLets kind body
+        bs' <- traverse (compileInBinding kind) bs
+        case r of
+            NonRec -> foldM (compileNonRecBinding kind) body' bs'
+            Rec    -> compileRecBindings kind body' bs'
+    Var x n -> pure $ Var x n
+    TyAbs x n k t -> TyAbs x n k <$> compileLets kind t
+    LamAbs x n ty t -> LamAbs x n ty <$> compileLets kind t
+    Apply x t1 t2 -> Apply x <$> compileLets kind t1 <*> compileLets kind t2
+    Constant x c -> pure $ Constant x c
+    Builtin x bi -> pure $ Builtin x bi
+    TyInst x t ty -> TyInst x <$> compileLets kind t <*> pure ty
+    Error x ty -> pure $ Error x ty
+    IWrap x tn ty t -> IWrap x tn ty <$> compileLets kind t
+    Unwrap x t -> Unwrap x <$> compileLets kind t
+
+compileInBinding
+    :: Compiling m e a
+    => LetKind
+    -> Binding TyName Name (Provenance a)
+    -> m (Binding TyName Name (Provenance a))
+compileInBinding kind b = case b of
+    TermBind x d rhs -> TermBind x d <$> compileLets kind rhs
+    _                -> pure b
+
+compileRecBindings :: Compiling m e a => LetKind -> PIRTerm a -> [Binding TyName Name (Provenance a)] -> m (PIRTerm a)
+compileRecBindings kind body bs =
+    let
+        partitionBindings = partition (\case { TermBind {} -> True ; _ -> False; })
+        (termBinds, typeBinds) = partitionBindings bs
+    in if null typeBinds then compileRecTermBindings kind body termBinds
+       else if null termBinds then compileRecTypeBindings kind body typeBinds
+       else ask >>= \p -> throwing _Error $ CompilationError p "Mixed term and type bindings in recursive let"
+
+compileRecTermBindings :: Compiling m e a => LetKind -> PIRTerm a -> [Binding TyName Name (Provenance a)] -> m (PIRTerm a)
+compileRecTermBindings RecTerms body bs = do
+    binds <- forM bs $ \case
+        TermBind _ vd rhs -> pure $ PIR.Def vd rhs
+        _ -> ask >>= \p -> throwing _Error $ CompilationError p "Internal error: type binding in term binding group"
+    compileRecTerms body binds
+compileRecTermBindings _ body bs = ask >>= \p -> pure $ Let p Rec bs body
+
+compileRecTypeBindings :: Compiling m e a => LetKind -> PIRTerm a -> [Binding TyName Name (Provenance a)] -> m (PIRTerm a)
+compileRecTypeBindings Types body bs = do
+    binds <- forM bs $ \case
+        DatatypeBind _ d -> pure d
+        _ -> ask >>= \p -> throwing _Error $ CompilationError p "Internal error: term or type binding in datatype binding group"
+    compileRecDatatypes body binds
+compileRecTypeBindings _ body bs = ask >>= \p -> pure $ Let p Rec bs body
+
+compileNonRecBinding :: Compiling m e a => LetKind -> PIRTerm a -> Binding TyName Name (Provenance a) -> m (PIRTerm a)
+compileNonRecBinding NonRecTerms body (TermBind x d rhs) = local (const x) $ local (TermBinding (varDeclNameString d)) $
+   PIR.mkTermLet <$> ask <*> pure (PIR.Def d rhs) <*> pure body
+compileNonRecBinding Types body (TypeBind x d rhs) = local (const x) $ local (TypeBinding (tyVarDeclNameString d)) $
```

Comment:
I'm a bit puzzled by those sequences of `local`s. Are you modifying the same environment twice discarding the result of the rightmost modification?

Reply:
The second modification builds on the first. This was less weird before I refactored it :sweat_smile: . Let me fix that.

Reply:
> The second modification builds on the first.

Oh I see where I was wrong, I thought it's `local f . local g ~ local (f . g)`, but it's `local f . local g ~ local (g . f)` instead. Thanks!